File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& h, int k) {
        vector<int> ans;
        int n=h.size();
        deque<int> q;
        for(int i=0;i<n;i++){
        while(!q.empty()&&h[q.back()]<=h[i]) q.pop_back();
        if(!q.empty()&&q.front()==i-k) q.pop_front();
        q.push_back(i);
        if(i>=k-1) ans.push_back(h[q.front()]);
        }
        return ans;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
import java.util.*;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int n = nums.length;
        int[] res = new int[n - k + 1];
        int j = 0;

        for (int i = 0; i < n; i++) {
            // Remove indices out of current window
            if (!dq.isEmpty() && dq.peekFirst() <= i - k) {
                dq.pollFirst();
            }

            // Remove smaller values as they are useless
            while (!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]) {
                dq.pollLast();
            }

            // Add current index
            dq.addLast(i);

            // Store result once we have a full window
            if (i >= k - 1) {
                res[j++] = nums[dq.peekFirst()];
            }
        }

        return res;
    }
}
